### PR TITLE
Rubber band zoom fix

### LINKED ISSUE #153: Zoom to Area in Plots
Zooming in plots over the horizontal axis only is not sufficient - this should be an area zoom.

### PR DESCRIPTION
**Pull request by jswift-stfc**
Initial fix for the rubber band zooming, to allow it to scale both axes. Closes #153.

### CODE REVIEW

**Review by trisyoungs — CHANGES_REQUESTED** [review]
Unfortunately, while the mechanism works, the area zoomed to doesn't represent that bounded visually by the rubberband.

If I draw a rectangle from x1,y1, to x2,y2 I notice that:
- The larger/higher value of y1 and y2 becomes the new maximum y limit of the graph, but the lower limit is "something else" (I haven't quite managed to work out what, but it certainly isn't the lower value of y1,y2.
- Similarly, the larger of x1,x2 becomes the *lower* limit of the zoomed region - the upper value again doesn't reflect the other of the two values.

**Comment by jswift-stfc** [comment]
I think I have fixed this, or made it better at least. Now I simply create a QRect: `QRect(x1, y1, x2-x1, y2-y1)` and normalise it, and then use the built-in zoom functionality, to zoom in on this region (this does not set the limits of the graph to that of the rect, but rather centers the rect and scales it, in a sense). Is this functionality okay?

**Comment by trisyoungs** [comment]
Current behaviour seems to be the same as before (in my previous comment)!

File: src/gui/widgets/gudpy_charts.py [hunk 1]
```diff
@@ -1012,6 +1019,42 @@ def copyPlot(self):
...
             event.accept()
+        elif event.button() == Qt.MouseButton.LeftButton:
+            # Catch the end of the rubber band.
+            self.rubberBandEnd = event.pos()
+
+            zoomFactor = 2.0
+
+            # Collect dims for rubber band rect.
+            top = max([self.rubberBandOrigin.x(), self.rubberBandEnd.x()])
+            left = min([self.rubberBandOrigin.y(), self.rubberBandEnd.y()])
```
[the remark this answers is not in the document]
Comment (jswift-stfc) [reply src/gui/widgets/gudpy_charts.py]: Yes, this was a mistake in the original version. The new version constructs the `QRect` to zoom into as such:

```
# Collect dims for rubber band rect.
width = event.pos().x() - self.rubberBandOrigin.x()
height = event.pos().y() - self.rubberBandOrigin.y()

# Create QRect area to zoom in on.
zoomArea = QRect(
    self.rubberBandOrigin.x(),
    self.rubberBandOrigin.y(),
    width,
    height
).normalized()
```
Where `rubberBandOrigin` is the position of the initial mouse click event.

**Comment by jswift-stfc** [comment]
> Current behaviour seems to be the same as before (in my previous comment)!

https://user-images.githubusercontent.com/87013792/149123331-42ffe1ec-f043-40f4-8eca-8aa1a7bc8835.mov


Is this the type of behaviour you see?

**Comment by trisyoungs** [comment]
Hmmm. Obviously a mix-up in downloaded artifacts somehow (or an old PR webpage).  Zoom does work correctly!

**Review by trisyoungs — APPROVED** [review]

**Comment by jswift-stfc** [comment]
Hurrah!

**Pull request merged**